### PR TITLE
Add BibTeX layer.

### PR DESCRIPTION
**Pull request by JP-Ellis**
Also integrates with org mode and LaTeX in order to manage citations.

Let me know if there anything you'd like me to change :)

### CODE REVIEW

**Comment by TheBB** [comment]
I think this should be in the latex layer. There's no need for it to be separate.

Also, please don't add keybindings from one layer to the README of another.

CC @justbur

**Comment by justbur** [comment]
Yeah it doesn't need to be separate. Also there's no reason dash, hydra and key-chord should be here.

**Comment by TheBB** [comment]
Oh, and don't `require` so many things in an init block please. Leave things to be loaded on demand whenever you can.

**Comment by JP-Ellis** [comment]
So, I've moved the LaTeX-mode and org-mode keybinding to be in the same layer (I wasn't sure which way was better, so that's no issue).

Regarding `dash`, `hydra` and `key-chord`: when I first tried to write this layer (back when `org-ref` wasn't on Melpa), `org-ref` wouldn't run at all unless these packages were loaded so that's the reason they are there.  I'll have a look at removing them and see if it all works.

As for the `require` in the init block, what do you suggest instead?  These are optional libraries provided in `org-ref` which are quite useful and they are not loaded otherwise.

Now, why do you think it should be integrated with the latex layer?  If anything, `org-ref` should be integrated into the `org` layer because of all the functionalities added to `org-mode`, but then I decided to not do that because really, this layer addresses the editing of BibTeX files, not org files.  Then of course, BibTeX files are pretty commonly used with LaTeX, but that alone I don't think is enough of a reason to integrate it into the LaTeX layer (what about people who use BibTeX with org-mode?).  You can see from the added keybinding that this new layer hardly affects org- and LaTeX-mode.

**Comment by TheBB** [comment]
Org-ref depends on dash, hydra and key-chord, as you can see here: http://melpa.org/#/org-ref So you shouldn't have to install those packages manually.

If you need to manually `require` something before it's used you are probably doing it wrong. I'm not familiar enough with org-ref and the setup you are trying to achieve, but I've written an introduction to how loading and autoloading works in Emacs [here](https://github.com/syl20bnr/spacemacs/blob/develop/doc/LAYERS.org) which might be helpful to you.

If there's a legitimate need for bibtex to be separate from latex I'm fine with that, but then the latex layer should depend on it, at least.

**Comment by justbur** [comment]
That's annoying that org-ref depends on hydra and key-chord.

**Comment by JP-Ellis** [comment]
Indeed, since `org-ref` has become available on melpa, the dependencies have been updated and are no longer necessary.  I have amended the commit to reflect that.

Now regarding the `require`, it seems like the functions need to have

```lisp
;;;###autoload
(defun foo () ...)
```

before they are declared, and currently `org-ref` does not do that at all... anywhere.  So at present, it seems like the `require` are necessary.  I will open an issue and/or pull request in `org-ref` later addressing that.

Finally, you mention that "the latex layer should depend on it, at least."; however, the LaTeX layer currently does not depend on BibTeX.  AucTeX (and it's extension RefTeX) offers some support to parse BibTeX files associated with a LaTeX file, but (afaik) don't provide any added functionalities within BibTeX files themselves and also don't add anything to org-mode.

**Comment by justbur** [comment]
Also keys should probably be set in init and there's no point to `(use-package x :defer t)` which expands to `nil`

**Comment by TheBB** [comment]
You can add additional autoloads with `use-package` (see same link as before, or the `use-package` documentation). E.g.

```lisp
(use-package org-ref
  :commands (some-func other-func)
  ;; ...
  )
```

will add autoloads for `some-func` and `other-func`.

**Comment by benquebec** [comment]
Wouldn't it be time to discuss what we'd like to have in the latex layer to better coordinate the effort?

Regarding this PR, it uses "o" as a binding which should be left for personal use.

**Comment by JP-Ellis** [comment]
@TheBB I tried doing that, but the problem is that `org-ref-arxiv` (for example), is not a package.  Trying to add

```lisp
(defun bibtex/init-org-ref-arxiv ()
  (use-package org-ref-arxiv
    :commands (...)
    :defer t))
```

results in Spacemacs trying to download `org-ref-arxiv` from MELPA.  Is there a way around that?  

Regarding the `o` binding... my apologies I wasn't aware this was left for personal use.  Is there another mnemonic for `open` that is commonly used in Spacemacs?

**Comment by TheBB** [comment]
@JP-Ellis Yes, you can do the `(use-package org-ref-arxiv ...)` inside `bibtex/init-org-ref`. Since `org-ref-arxiv` is a *feature* in the `org-ref` *package* that would be fine in my book.

**Comment by JP-Ellis** [comment]
I have removed the various `require` and at this stage only those functions explicitly used in the key bindings are added to the autoload.

I'm still not sure what the favoured mnemonic alternative for `open` is...

**Comment by justbur** [comment]
Why not use the single letters p, n, and b for the open commands?

File: layers/+lang/bibtex/README.org [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+| Key Binding | Description     |
+|-------------+-----------------|
+| ~SPC m i c~ | Insert citation |
+
+and LaTeX mode also gains:
+
+| Key Binding | Description            |
+|-------------+------------------------|
+| ~SPC m i i~ | insert bibtex citation |
```
Comment (cpaulik) [inline layers/+lang/bibtex/README.org]: I guess this should also be `SPC m i c` ?

**Comment by JP-Ellis** [comment]
I have update the typo spotted by @cpaulik, and changed the various open keybinding to just `b`, `n` and `p` instead of `ob`, `on` and `op`.

**Comment by benquebec** [comment]
Might it be useful to add https://github.com/tmalsburg/helm-bibtex to this layer?

**Comment by JP-Ellis** [comment]
Org ref uses `helm-bibtex` as a dependency already.  Is there anything in particular you wanted to add from that package?  I can add them to this pull request before it is merged.

**Comment by benquebec** [comment]
Sorry, I did not know. I don't use org hence my suggestion to include helm-bibtex.

**Comment by JP-Ellis** [comment]
I'll have a look at any feature from `helm-bibtex` and add shortcuts for any I find interesting. (I can't do this for another few days though... So please don't merge it quite yet then).

**Comment by robbyoconnor** [comment]
This belongs in the `latex` layer -- I concur with @TheBB

**Comment by cpaulik** [comment]
@robbyoconnor I don't think that it must be in the `latex` layer. I use `org-ref` and `helm-bibtex` a lot outside of Latex.

**Comment by robbyoconnor** [comment]
Why not have an option -- options are okay. Why shouldn't this be in the `latex` layer -- We can have an option of which to use.

**Comment by JP-Ellis** [comment]
The reason I think this should not be in the latex layer is that BibTeX files are used in more contexts than just LaTeX documents, in particular, org and markdown both support BibTeX references, and I wouldn't be surprised if other text markups (such as reStructuredText) support it too.

In addition, each of these mode depend in BibTeX just as much as any other (that is, hardly at all and only to insert citations).  This then begs the question: why should BibTeX be part of the latex layer?  Why not the org layer? or markdown layer?

**Comment by robbyoconnor** [comment]
I guess I could buy that argument -- ultimately it's @syl20bnr's opinion that matters...I doubt there'd be a revolt in either case...

**Comment by benquebec** [comment]
My guess is someone new to spacemacs is more likely to look for bibtex in the latex layer which contains everything TeX. Based on conversations on gitter, that might ease support...

**Comment by JP-Ellis** [comment]
Well, if that's the worry I can amend this pull request so that the LaTeX layer mentions this new layer for those who want better BibTeX support.

**Comment by TheBB** [comment]
This isn't a problem... you can make the LaTeX layer depend on the BibTeX layer.

**Comment by robbyoconnor** [comment]
I like the solution that @TheBB and @JP-Ellis suggested.

**Comment by JP-Ellis** [comment]
I have made a few amendments to the pull request.  Firstly, I mention the BibTeX layer in the org, markdown and LaTeX layers so that new users can more easily find how to manage citations and references in each of these modes (let me know if there are other modes that should be added).  I also added a key binding to Markdown mode.

Let me know if there is anything else that should be changed in this pull request :)

File: layers/org/README.org [hunk 2]
```diff
@@ -30,6 +30,9 @@
 * Description
 This layer enables  [[http://orgmode.org/][org mode]] for Spacemacs.
 
+For more extensive support of references through BibTeX files, have a look at
+the [[./+lang/bibtex][BibTeX layer]].
+
```
Comment (robbyoconnor) [inline layers/org/README.org]: Is this the right path? You mean the following right?

```org 
 [[../+lang/bibtex]]
```

Comment (JP-Ellis) [inline layers/org/README.org]: Indeed, there is a missing `.`, I'll update the pull request.

**Comment by JP-Ellis** [comment]
So, what's the status on this pull request, is there anything else that needs to be modified?

Also, I use helm, but recently spacemacs started supporting ivy.  Does this affect this layer at all as it depends on helm-bibtex?

**Comment by StreakyCobra** [comment]
@syl20bnr has to go over each PRs that concerns a new layer, so you probably have to be patient :-)

You don't have to care about ivy for now.

**Comment by JP-Ellis** [comment]
Ok, all good, I'm happy to wait :)  I just wanted to make sure that you guys weren't waiting on something more.

**Comment by StreakyCobra** [comment]
@JP-Ellis Maybe people participating in this PR still have remarks though, I was speaking more generally ;-)

**Comment by JP-Ellis** [comment]
I have just updated the pull request.  The main changes are:

- make use of the `packages-config.el`;
- Have a better README file, including logo and a better description of BibTeX and this layer;
- Remove the explicit autoloading of functions through `use-package` thanks to jkitchin/org-ref#117.

**Comment by StreakyCobra** [comment]
:green_heart: @syl20bnr We have a nice candidate for merge here ;-)

**Comment by DanTheColoradan** [comment]
@JP-Ellis, I tried inserting the bibtex layer to +lang manually. I added bibtex to my layer config and as far as I could tell, it loaded A-ok. Then, I opened a `.md` file and the `SPC m i c` feature doesn't come up.

My spacemacs is up to date. Perhaps I need to use a develop version?

**Comment by JP-Ellis** [comment]
@DanTheColoradan That's strange... Unless I introduced a minor regression in the latest changes?  Just making sure, you have the markdown layer enabled too?  I'll have a look into it.

**Comment by DanTheColoradan** [comment]
Yup. Markdown layer is enabled. I need to figure out how to debug layers...

**Comment by JP-Ellis** [comment]
@DanTheColoradan I think you do have to use the `develop` branch in order to use this layer since there recently was a change to the way packages are loaded (aa4b6e68618ebd99e93c4093993799adea70ca1a).

**Comment by robbyoconnor** [comment]
Did you restart spacemacs? Sometimes you need to restart spacemacs...

**Comment by robbyoconnor** [comment]
In addition to that -- the new structure is currently only in `develop` -- As much as I always say it...if you wanna play with this...use `develop` :grin:

**Comment by DanTheColoradan** [comment]
K :D Either I'll wait, or it'll be bleeding edge for me! Thanks and looking forward to a successful merge :)

**Comment by DanTheColoradan** [comment]
@JP-Ellis, can you add a little more documentation?

I was attempting to set my bibliography location using org-ref, but it looks like the correct way it to use helm-bibtex, correct? Perhaps under the Installation header, to see [helm-bibtex](https://github.com/tmalsburg/helm-bibtex) for configuration options?

**Comment by JP-Ellis** [comment]
@DanTheColoradan Thanks for the suggestion.  I just amended the README to have a Configuration section :)

**Comment by DanTheColoradan** [comment]
Nice. :beers:

File: layers/+lang/bibtex/packages-config.el [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    "ic" 'org-ref-helm-insert-cite-link))
+
+(defun bibtex/post-init-org ()
+  (spacemacs/set-leader-keys-for-major-mode 'org-mode
+    "ic" 'org-ref-helm-insert-cite-link))
+
+(defun bibtex/post-init-markdown-mode ()
+  (spacemacs/set-leader-keys-for-major-mode 'markdown-mode
+    "ic" 'org-ref-helm-insert-cite-link))
+
```
Comment (robbyoconnor) [inline layers/+lang/bibtex/packages-config.el]: This could probably done with a in a DRYER way...

Comment (JP-Ellis) [inline layers/+lang/bibtex/packages-config.el]: At this stage, it's only three modes which need this, so I thought that repeating the code is not too bad, especially given that it's one command for each mode.  If, in the future, many more modes need a shortcut added then we can always use a DRYer way.

Having said that, if you want to do a pull request against my fork then I'll include it (though I'll probably squash it so as to have only 1 commit).

Comment (robbyoconnor) [inline layers/+lang/bibtex/packages-config.el]: I'm not sure how to do it -- there has to be a way to do this in a DRY way...but I have stuff I need to do -- it's a mere suggestion...ultimately @syl20bnr is the boss here.

**Comment by michelk** [comment]
+1 for a BibTeX-Layer

**Comment by JP-Ellis** [comment]
I rebased the commit on top of all the changes so that it can be merged again.  Let me know what else is needed for this pull request (or whether I should close it).

**Comment by StreakyCobra** [comment]
@JP-Ellis You probably have to wait, @syl20bnr is merging new layer PR in person, so it may take time.

**Comment by cpaulik** [comment]
@JP-Ellis Have you seen the new features for adding references shown in https://www.youtube.com/watch?v=2t925KRBbFc ?

I think those do not work out of the box since they are in `org-ref-pdf.el`

**Comment by JP-Ellis** [comment]
@cpaulik I haven't seen these new features, I'll have a look at updating the pull request.  This will have to be later next week though (probably next weekend).

**Comment by AlejandroCatalina** [comment]
I just saw [biblio.el](https://github.com/cpitclaudel/biblio.el) package and seems a good candidate to be included in this layer, unless i'm missing some package that already has that functionality and is already contained (org-ref maybe?).

We can also leave the layer as it is and include it in a later PR, which i'm highly willing to do ;)

**Comment by JP-Ellis** [comment]
@AlejandroCatalina That looks like a pretty good package as well, though some features do seem to overlap with those provides by org-ref.  It seems like @syl20bnr is merging this layer in (yay!), so perhaps it would be best to include `biblio.el` in a later pull request.

Also, @cpaulik, I haven't had time to update the pull request with the features from that video.  I'll create another pull request some other time perhaps, or you can have a look at doing so.

Lastly, I see that there are now conflicts with the branch, so I'll quickly rebase it.

**Comment by AlejandroCatalina** [comment]
Totally fine @JP-Ellis , i'll PR it later then :+1:

**Comment by syl20bnr** [comment]
Very nice! Thank you ! 👍 
I made some modifications in commit https://github.com/syl20bnr/spacemacs/commit/9160109
> Remove packages-config.el
Fix format of README

`packages-config.el` was introduced in develop branch but we decided to remove it in the end.

Cherry-picked into develop branch, you can safely delete your branch.

**Comment by yilihong** [comment]
any progress on this? thank you.

**Comment by JP-Ellis** [comment]
@yilihong The pull request has been merged and is now in the  [develop branch](https://github.com/syl20bnr/spacemacs/tree/develop/layers/+lang), though now yet in the [master branch](https://github.com/syl20bnr/spacemacs/tree/master/layers/+lang).  I don't know how long it takes for new layers to migrate from the develop branch to the master branch though... maybe someone else knows?  If you want to use this now, you'll have to switch to the development branch.

**Comment by yilihong** [comment]
@JP-Ellis maybe within the next few releases?

I got it org-ref to work through some hacks, but it will be wonderful to have a formal layer for it.

**Comment by robbyoconnor** [comment]
@yilihong you can join the dark side and use `develop` -- we have :cookie:'s  -- lots of them!

**Pull request closed without merging**